### PR TITLE
Adds ruff linter, fixes unused variables errors

### LINKED ISSUE #8665: Start to use Pylint
Issues as https://github.com/DefectDojo/django-DefectDojo/issues/8636#issuecomment-1719081247 (fixed in https://github.com/DefectDojo/django-DefectDojo/pull/8663) would not happen if we would have a track of unused variables.

I suggest using Pylint to keep the code clean. Flake8 does a lot but it is not enough.

### PR DESCRIPTION
**Pull request by dwalleck**
**Description**

closes #8665 

This change adds the ruff linter, which is a Rust-driven Python tool for timely linting. It covers the same ruleset as flake8 and other Python linting tools. In particular, this change adds a rule to disallow unused variables. Once we can guarantee that ruff is covering everything that flake8 covers, we should be able to remove flake8 from gating.

**Test results**

The ruff tool reports no errors with the specified ruleset.

### CODE REVIEW

**Comment by mtesauro** [comment]
@dwalleck Thanks for this contribution.

This looks very promising.  Please get those couple of failing tests to pass so we can merge this. :+1:

**Comment by kiblik** [comment]
Wow, nice. I suppose it identified many places which will optimise evaluation.

Originally I was scared that DD is not validating passwords on change but I found out it is done in the `form.is_valid()`

I suppose this should not be simply merged because in some cases it will need the removal of the whole line (not only `var =`) and in some other cases usage of data should be added (not only the removal of their processing).

File: dojo/cred/views.py [hunk 1]
```diff
@@ -326,7 +326,7 @@ def view_cred_finding(request, fid, ttid):
         title="Credential Manager", top_level=False, request=request)
     cred_type = "Finding"
     edit_link = None
-    view_link = reverse(
+    reverse(
         'view_cred_finding', args=(
             fid,
             cred.id,
```
Comment (Maffooch) [inline dojo/cred/views.py:332]: Agreed with Kiblik ^ here is an example of still doing the unused processing

Comment (dwalleck) [reply dojo/cred/views.py:332]: Good catch. I'll go back over these changes a bit more carefully

File: dojo/cred/views.py [hunk 2]
```diff
@@ -270,7 +265,7 @@ def view_cred_engagement_test(request, tid, ttid):
         title="Credential Manager", top_level=False, request=request)
     cred_type = "Test"
     edit_link = None
-    view_link = reverse(
+    reverse(
```
Comment (Maffooch) [inline dojo/cred/views.py:268]: This function could be removed

File: dojo/tools/outpost24/parser.py [hunk 3]
```diff
@@ -57,9 +57,6 @@ def get_findings(self, file, test):
                 else:
                     severity = "Critical"
             cvss_description = detail.findtext("cvss_vector_description")
-            cvss_vector = detail.findtext("cvss_v3_vector") or detail.findtext(
-                "cvss_vector"
-            )
```
Comment (Maffooch) [inline dojo/tools/outpost24/parser.py:62]: This one could be worth keeping to improve the parser

**Review by Maffooch — CHANGES_REQUESTED** [review]
Did a quick sweep

File: dojo/group/utils.py [hunk 4]
```diff
@@ -15,7 +15,7 @@ def get_auth_group_name(group, attempt=0):
         auth_group_name = group.name + '_' + str(attempt)
 
     try:
-        auth_group = Group.objects.get(name=auth_group_name)
+        _ = Group.objects.get(name=auth_group_name)
```
[the remark this answers is not in the document]
Comment (kiblik) [reply dojo/group/utils.py:18]:
```suggestion
        Group.objects.get(name=auth_group_name)
```

File: dojo/jira_link/views.py [hunk 5]
```diff
@@ -326,7 +326,7 @@ def new_jira(request):
             jira_password = jform.cleaned_data.get('password')
 
             logger.debug('calling get_jira_connection_raw')
-            jira = jira_helper.get_jira_connection_raw(jira_server, jira_username, jira_password)
+            _ = jira_helper.get_jira_connection_raw(jira_server, jira_username, jira_password)
```
Comment (kiblik) [inline dojo/jira_link/views.py:329]:
```suggestion
            jira_helper.get_jira_connection_raw(jira_server, jira_username, jira_password)
```

File: dojo/management/commands/migrate_surveys.py [hunk 6]
```diff
@@ -39,7 +39,7 @@ def handle(self, *args, **options):
             # Get unique ploymorphic id for the system
             ctype_id = 0
             # First create a temp question to pull the polymorphic_ctype_id from
-            created_question = TextQuestion.objects.create(optional=False, order=1, text='What is love?')
+            _ = TextQuestion.objects.create(optional=False, order=1, text='What is love?')
```
Comment (kiblik) [inline dojo/management/commands/migrate_surveys.py:42]:
```suggestion
            TextQuestion.objects.create(optional=False, order=1, text='What is love?')
```

File: dojo/metrics/views.py [hunk 7]
```diff
@@ -390,7 +390,7 @@ def metrics(request, mtype):
         request.GET._mutable = True
         request.GET.appendlist('test__engagement__product__prod_type', mtype)
         request.GET._mutable = False
-        product = pt[0].name
+        pt[0].name
```
[the remark this answers is not in the document]
Comment (kiblik) [reply dojo/metrics/views.py:393]:
```suggestion
```

File: dojo/forms.py [hunk 8]
```diff
@@ -3012,7 +3011,7 @@ def __init__(self, *args, **kwargs):
             initial=initial_answer,
         )
 
-        answer = self.fields['answer']
+        self.fields['answer']
```
[the remark this answers is not in the document]
Comment (kiblik) [reply dojo/forms.py:3014]:
```suggestion
```

File: dojo/models.py [hunk 9]
```diff
@@ -2821,7 +2821,7 @@ def github(self):
 
     def has_github_issue(self):
         try:
-            issue = self.github_issue
+            _ = self.has_github_issue
```
[the remark this answers is not in the document]
Comment (kiblik) [reply dojo/models.py:2824]:
```suggestion
            self.github_issue
```

File: dojo/forms.py [hunk 10]
```diff
@@ -661,7 +660,7 @@ class MergeFindings(forms.ModelForm):
         help_text="The action to take on the merged finding. Set the findings to inactive or delete the findings.")
 
     def __init__(self, *args, **kwargs):
-        finding = kwargs.pop('finding')
+        _ = kwargs.pop('finding')
```
Comment (kiblik) [inline dojo/forms.py:663]:
```suggestion
        kwargs.pop('finding')
```

File: dojo/jira_link/helper.py [hunk 11]
```diff
@@ -786,7 +785,7 @@ def add_jira_issue(obj, *args, **kwargs):
         j_issue.jira_creation = timezone.now()
         j_issue.jira_change = timezone.now()
         j_issue.save()
-        issue = jira.issue(new_issue.id)
+        _ = jira.issue(new_issue.id)
```
Comment (kiblik) [inline dojo/jira_link/helper.py:788]:
```suggestion
        jira.issue(new_issue.id)
```

File: dojo/product/views.py [hunk 12]
```diff
@@ -61,11 +61,10 @@
 
 def product(request):
     # validate prod_type param
-    product_type = None
     if 'prod_type' in request.GET:
         p = request.GET.getlist('prod_type', [])
         if len(p) == 1:
-            product_type = get_object_or_404(Product_Type, id=p[0])
+            _ = get_object_or_404(Product_Type, id=p[0])
```
Comment (kiblik) [inline dojo/product/views.py:67]:
```suggestion
            get_object_or_404(Product_Type, id=p[0])
```

File: dojo/product/views.py [hunk 13]
```diff
@@ -1001,16 +994,14 @@ def delete_product(request, pid):
...
     jira_project_form = None
     jira_epic_form = None
 
     product = Product.objects.get(id=pid)
-    jira_error = False
 
     if request.method == 'POST':
         form = EngForm(request.POST, cicd=cicd, product=product, user=request.user)
-        jira_project = jira_helper.get_jira_project(product)
+        _ = jira_helper.get_jira_project(product)
```
Comment (kiblik) [inline dojo/product/views.py:1004]:
```suggestion
        jira_helper.get_jira_project(product)
```

File: dojo/product/views.py [hunk 14]
```diff
@@ -1069,7 +1060,7 @@ def new_eng_for_app(request, pid, cicd=False):
                        product=product, user=request.user)
 
         if get_system_setting('enable_jira'):
-            jira_project = jira_helper.get_jira_project(product)
+            _ = jira_helper.get_jira_project(product)
```
Comment (kiblik) [inline dojo/product/views.py:1063]:
```suggestion
            jira_helper.get_jira_project(product)
```

File: dojo/reports/widgets.py [hunk 15]
```diff
@@ -28,7 +28,7 @@ class CustomReportJsonForm(forms.Form):
     def clean_json(self):
         jdata = self.cleaned_data['json']
         try:
-            json_data = json.loads(jdata)
+            _ = json.loads(jdata)
```
Comment (kiblik) [inline dojo/reports/widgets.py:31]:
```suggestion
            json.loads(jdata)
```

File: dojo/user/views.py [hunk 16]
```diff
@@ -181,7 +181,7 @@ def delete_alerts(request):
     alerts = Alerts.objects.filter(user_id=request.user)
 
     if request.method == 'POST':
-        removed_alerts = request.POST.getlist('alert_select')
+        _ = request.POST.getlist('alert_select')
```
Comment (kiblik) [inline dojo/user/views.py:184]:
```suggestion
        request.POST.getlist('alert_select')
```

**Review by kiblik — CHANGES_REQUESTED** [review]
Some feedback from my review

**Comment by kiblik** [comment]
Is it possible to approve the workflow here, please? It would be good to know, what kind of damage it creates in tests and unittests

File: .github/workflows/ruff.yml [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+      - name: Checkout
+        # for non PR runs we just checkout the default, which is a sha on a branch probably
+        if: github.event_name != 'pull_request' && github.event_name != 'pull_request_target'
+        uses: actions/checkout@v4
+      
+      - name: Install dev dependencies
+        run: pip install -r requirements-dev.txt
+
+      - name: Run ruff
+        run: ruff dojo
```
[the remark this answers is not in the document]
Comment (dwalleck) [reply .github/workflows/ruff.yml:37]: Those are both great ideas! However, for the sake of not making this PR even larger, I would prefer to add the linting of those directories in a separate PR that quickly follows this one.

File: dojo/api_v2/views.py [hunk 18]
```diff
@@ -597,7 +597,7 @@ def notes(self, request, pk=None):
             serialized_note = serializers.NoteSerializer(
                 {"author": author, "entry": entry, "private": private}
             )
-            result = serializers.EngagementToNotesSerializer(
+            serializers.EngagementToNotesSerializer(
```
[the remark this answers is not in the document]
Comment (dwalleck) [reply dojo/api_v2/views.py:600]: Thanks! I will remove it

File: dojo/cred/views.py (hunk 2 above)
Comment (dwalleck) [reply dojo/cred/views.py:268]: Thanks. I'll remove it

File: requirements-dev.txt [hunk 19]
```diff
@@ -0,0 +1 @@
+ruff==0.0.290
```
[the remark this answers is not in the document]
Comment (dwalleck) [reply requirements-dev.txt:1]: It's very common to have a requirements-dev/test/etc for tools used only in developing an app. Right now there is only one dependency, but there will be many more.

File: .github/workflows/ruff.yml (hunk 17 above)
[the remark this answers is not in the document]
Comment (kiblik) [reply .github/workflows/ruff.yml:37]: Fully understand. No problem. It is a good idea.

File: requirements-dev.txt (hunk 19 above)
[the remark this answers is not in the document]
Comment (kiblik) [reply requirements-dev.txt:1]: Ok, no problem

**Review by mtesauro — APPROVED** [review]
Approved.

**Comment by mtesauro** [comment]
@dwalleck I approved assuming you've resolved the pending reviews - the people who put those in should have a look and close those out soon.

@kiblik @Maffooch

File: dojo/group/utils.py [hunk 20]
```diff
@@ -15,7 +15,7 @@ def get_auth_group_name(group, attempt=0):
         auth_group_name = group.name + '_' + str(attempt)
 
     try:
-        auth_group = Group.objects.get(name=auth_group_name)
+        Group.objects.get(name=auth_group_name)
```
Comment (Maffooch) [inline dojo/group/utils.py:18]: This can be removed now

**Review by Maffooch — CHANGES_REQUESTED** [review]
Commenting on the whole review instead of each instance. For instances of removing an unused variable, please also check out what the function is doing to determine if it can also be removed. In one instance I found, there are things like this:
Before:
```
engagements = Engagement.objects.filter(
             product__prod_type=product_type, test__finding__in=findings.qs
         ).distinct()
```
After:
```
Engagement.objects.filter(
             product__prod_type=product_type, test__finding__in=findings.qs
         ).distinct()
```
Though this filter call does not actually result in a DB query due to lazy eval by django, this is now just dead code that serves no purpose

**Review by kiblik — CHANGES_REQUESTED** [review]
Sorry guys, I prepared my review but forgot to submit it.

**Pull request closed without merging**